crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,359 @@
+        loop {
+            if !self.end_of_epoch {
+                self.schedule_synced_checkpoints(&mut finished)
+                    .await
+                    .unwrap_or_else(|err| {
+                        self.metrics.checkpoint_exec_errors.inc();
+                        panic!(
+                            "Failed to schedule synced checkpoints for execution: {:?}",
+                            err
+                        );
+                    });
+            }
+
+            tokio::select! {
+                // Check for completed workers and ratchet the highest_checkpoint_executed
+                // watermark accordingly. Note that given that checkpoints are guaranteed to
+                // be processed (added to FuturesOrdered) in seq_number order, using FuturesOrdered
+                // guarantees that we will also ratchet the watermarks in order.
+                Some(Ok((checkpoint, next_committee))) = finished.next() => {
+                    match next_committee {
+                        None => {
+                            self.metrics.last_executed_checkpoint.set(checkpoint.sequence_number() as i64);
+                            self.checkpoint_store
+                                .update_highest_executed_checkpoint(&checkpoint)
+                                .unwrap();
+                        }
+                        // Last checkpoint of epoch -- Note that we must not update the
+                        // highest executed watermark until after reconfig has completed!
+                        Some(committee) => return Some((checkpoint, committee)),
+                    }
+                }
+                // Check for newly synced checkpoints from StateSync.
+                received = self.mailbox.recv() => match received {
+                    Ok(checkpoint) => self.latest_synced_checkpoint = Some(checkpoint),
+                    // In this case, messages in the mailbox have been overwritten
+                    // as a result of lagging too far behind. In this case, we need to
+                    // nullify self.latest_synced_checkpoint as the latest synced needs to
+                    // be read from the watermark
+                    Err(RecvError::Lagged(num_skipped)) => {
+                        self.latest_synced_checkpoint = None;
+
+                        warn!(
+                            "Checkpoint Execution Recv channel overflowed {:?} messages",
+                            num_skipped,
+                        );
+                        self.metrics
+                            .checkpoint_exec_recv_channel_overflow
+                            .inc_by(num_skipped);
+                    }
+                    Err(RecvError::Closed) => {
+                        info!("Checkpoint Execution Sender (StateSync) closed channel");
+                        return None;
+                    }
+                },
+            }
+        }
+    }
+
+    async fn schedule_synced_checkpoints(
```

Comment:
This function doesn't need to be `async` once you change `schedule_checkpoint` to also not need to be async.